File: app/src/main/java/info/muge/apkextractor/tasks/SearchAppItemTask.java
package info.muge.apkextractor.tasks;

import androidx.annotation.NonNull;

import info.muge.apkextractor.Global;
import info.muge.apkextractor.items.AppItem;
import info.muge.apkextractor.utils.PinyinUtil;

import java.util.ArrayList;
import java.util.List;

public class SearchAppItemTask extends Thread {

    private volatile boolean isInterrupted=false;
    private final String search_info;
    private final List<AppItem> appItemList=new ArrayList<>();
    private final ArrayList<AppItem> result_appItems=new ArrayList<>();
    private final SearchTaskCompletedCallback callback;


    public SearchAppItemTask(List<AppItem>appItems,@NonNull String info,@NonNull SearchTaskCompletedCallback callback){
        this.search_info=info.trim().toLowerCase();
        this.appItemList.addAll(appItems);
        this.callback=callback;
    }

    @Override
    public void run() {
        super.run();
        for(AppItem item: appItemList){
            if(isInterrupted){
                break;
            }
            try{
                boolean b=(getFormatString(item.getAppName()).contains(search_info)
                        ||getFormatString(item.getPackageName()).contains(search_info)
                        ||getFormatString(item.getVersionName()).contains(search_info)
                        ||getFormatString(PinyinUtil.getFirstSpell(item.getAppName())).contains(search_info)
                        ||getFormatString(PinyinUtil.getFullSpell(item.getAppName())).contains(search_info)
                        ||getFormatString(PinyinUtil.getPinYin(item.getAppName())).contains(search_info))&&!search_info.trim().equals("");
                if(b) result_appItems.add(item);
            }catch (Exception e){e.printStackTrace();}
        }
        Global.handler.post(new Runnable() {
            @Override
            public void run() {
                if(!isInterrupted)callback.onSearchTaskCompleted(result_appItems,search_info);
            }
        });
    }

    public void setInterrupted(){
        isInterrupted=true;
    }

    private String getFormatString(@NonNull String s){
        return s.trim().toLowerCase();
    }

    public interface SearchTaskCompletedCallback{
        void onSearchTaskCompleted(@NonNull List<AppItem> appItems,@NonNull String keyword);
    }
}

File: app/src/main/java/info/muge/appshare/tasks/SearchAppItemTask.kt
package info.muge.appshare.tasks

import info.muge.appshare.items.AppItem
import info.muge.appshare.utils.PinyinUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext

/**
 * 搜索应用项任务（协程版）
 */
class SearchAppItemTask(
    appItems: List<AppItem>,
    info: String
) {
    private val searchInfo: String = info.trim().lowercase()
    private val appItemList = ArrayList<AppItem>(appItems)

    /**
     * 执行搜索（挂起函数，在 IO 线程执行）
     * 通过 Job.cancel() 取消
     */
    suspend fun execute(): List<AppItem> = withContext(Dispatchers.IO) {
        val resultItems = ArrayList<AppItem>()

        for (item in appItemList) {
            ensureActive() // 替代 isInterrupted 检查

            try {
                val matched = (getFormatString(item.getAppName()).contains(searchInfo) ||
                        getFormatString(item.getPackageName()).contains(searchInfo) ||
                        getFormatString(item.getVersionName()).contains(searchInfo) ||
                        getFormatString(PinyinUtil.getFirstSpell(item.getAppName())).contains(searchInfo) ||
                        getFormatString(PinyinUtil.getFullSpell(item.getAppName())).contains(searchInfo) ||
                        getFormatString(PinyinUtil.getPinYin(item.getAppName())).contains(searchInfo)) &&
                        searchInfo.trim().isNotEmpty()

                if (matched) {
                    resultItems.add(item)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        resultItems
    }

    private fun getFormatString(s: String): String {
        return s.trim().lowercase()
    }

    /**
     * 搜索关键词
     */
    val keyword: String get() = searchInfo

    /**
     * 搜索任务完成回调（保留向后兼容）
     */
    interface SearchTaskCompletedCallback {
        fun onSearchTaskCompleted(appItems: List<AppItem>, keyword: String)
    }
}
